controllers/k8ssandra/datacenters_test.go

```diff
@@ -0,0 +1,139 @@
+package k8ssandra
+
+import (
+	"context"
+	"testing"
+
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	api "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/test/framework"
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	analyticsWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				AnalyticsEnabled: true,
+			},
+		},
+	}
+
+	searchWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				SearchEnabled: true,
+			},
+		},
+	}
+
+	graphWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				GraphEnabled: true,
+			},
+		},
+	}
+
+	mixedWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				GraphEnabled:     true,
+				AnalyticsEnabled: true,
+			},
+		},
+	}
+
+	cassandraWorkloadDc = api.CassandraDatacenterTemplate{
+		DatacenterOptions: api.DatacenterOptions{
+			ServerVersion: "6.8.17",
+			DseWorkloads: &cassdcapi.DseWorkloads{
+				GraphEnabled:     false,
+				AnalyticsEnabled: false,
+				SearchEnabled:    false,
+			},
+		},
+	}
+)
+
+func datacentersTest(t *testing.T, ctx context.Context, f *framework.Framework, namespace string) {
```

Comment:
These are actually unit tests, not integration tests (in the sense of: tests using the testEnv framework). You should change the signature here to:

```go
func TestDatacenters(t *testing.T) {
	t.Run("DcUpgradePriorityTest", dcUpgradePriorityTest)
	t.Run("SortDatacentersForUpgradeTest", sortDatacentersForUpgradeTest)
	t.Run("SortNoChangeTest", sortNoChangeTest)
}
```

The advantage is that you can run them without spinning up an API server for nothing and so it's going to be much faster.